File: my-folder/0978-valid-mountain-array/solution.c


bool validMountainArray(int* arr, int arrSize){
    bool peakFound = false;
    if (!arr || (arrSize < 3) || (arr[0] > arr[1]))
        return false;
    for (size_t i = 1; i < arrSize; i++){
        if (arr[i] == arr[i-1])
            return false;
        if (!peakFound && (arr[i] < arr[i-1]))
            peakFound = true;
        else if (peakFound && (arr[i] > arr[i-1]))
            return false;
    }
    return peakFound;
}

File: my-folder/0978-valid-mountain-array/solution.rs
impl Solution {
    pub fn valid_mountain_array(arr: Vec<i32>) -> bool {
        use std::cmp::Ordering;

        if arr.len() < 3 {
            return false;
        }

        let mut comparisons = arr.windows(2).map(|pair| pair[0].cmp(&pair[1]));

        // Check for the initial increasing phase: the first comparison must be `Ordering::Less`
        if comparisons.next() != Some(Ordering::Less) {
            return false;
        }

        // Check the middle phase: transitions from `Less` to `Greater`, indicating a peak
        if !comparisons.by_ref().skip_while(|&x| x == Ordering::Less).next().eq(&Some(Ordering::Greater)) {
            return false;
        }

        // Verify the remaining phase: all comparisons should now be `Ordering::Greater`
        if comparisons.skip_while(|&x| x == Ordering::Greater).next().is_some() {
            return false;
        }

        true
    }
}
